Fix push to unqualified destination

jsoriano opened the pull request:
When pushing a revision to a not existing reference git cannot know what the reference is and it fails, for example the test added in this PR fails with:
```
E           error: unable to push to unqualified destination: notexisting
E           The destination refspec neither matches an existing ref on the remote nor
E           begins with refs/, and we are unable to guess a prefix based on the source ref.
E           error: failed to push some refs to '/tmp/tmpWe7qX3/main_bare'
```

The fix here assumes that we want to push to a new branch (we could want to push to a new tag).

Alternativelly we could force the client to be explicit about the kind of reference it wants to create (i.e. when pushing it should use the parameter with `ref/whatever` preffix, e.g: `repo.push(rev=fe1464e022f28c, ref_name='refs/heads/branch')` or `repo.push(rev=fe1464e022f28c, ref_name='refs/tags/branch')`, but this would require changes in clients, and wouldn't be scm-independent.

mrodm commented:
Right now, repoman is used to push tags through this `push` method using both parameters "ref_name" and "rev". If it is changed, it would create a branch named as the tag, wouldn't it ?

An example of use:

```        repository.update(branch)
        tagged_revision = repository.tag(tag, revision=changeset)
        self.logger.info("Tagged revision %s as %s" % (changeset, tag))
        repository.push(rev=tagged_revision,
                        dest=repository_url,
                        ref_name=tag)
```

jsoriano commented:
I think that @oscarsj saw a case in which it was expected to create a branch in the remote.
What I can do is to check if the ref exists, if it does I leave it as is, and it if doesn't I preffix it with `refs/heads`.

oscarsj commented:
I have been working with a quick patch that does exactly what jaime suggests, and seems to be working, so i'd go for that

jsoriano commented:
I have added the check to don't add the full ref path if it is an existing tag or branch. I have also added support to force the push, as suggested by @mrodm too.
